src/server/stream_family.cc

```diff
@@ -1364,6 +1502,124 @@ void StreamFamily::XAdd(CmdArgList args, ConnectionContext* cntx) {
+          break;
+        }
+        return ids;
+      }
+      ids.push_back(parsed_id.val);
+    }
+    args.remove_prefix(i);
+    return ids;
+  };
+
+  auto ids = GetXclaimIds();
+  if (!ids.size()) {
+    // No ids given.
+    return (*cntx)->SendError(kInvalidStreamId, kSyntaxErrType);
+  }
+
+  // parse the options
+  auto ParseXclaimOptions = [&]() {
+    size_t i = 0;
+    for (i = 0; i < args.size(); ++i) {
+      ToUpper(&args[i]);
+      string_view arg = ArgS(args, i);
+      bool remaining_args = args.size() - i - 1 > 0;
+
+      if (remaining_args) {
+        if (arg == "IDLE") {
+          arg = ArgS(args, ++i);
+          if (!absl::SimpleAtoi(arg, &opts.deliverytime)) {
+            return (*cntx)->SendError(kInvalidIntErr);
+          }
+          continue;
+        } else if (arg == "TIME") {
+          arg = ArgS(args, ++i);
+          if (!absl::SimpleAtoi(arg, &opts.deliverytime)) {
+            return (*cntx)->SendError(kInvalidIntErr);
+          }
+          continue;
+        } else if (arg == "RETRYCOUNT") {
+          arg = ArgS(args, ++i);
+          if (!absl::SimpleAtoi(arg, &opts.retry)) {
+            return (*cntx)->SendError(kInvalidIntErr);
+          }
+          continue;
+        } else if (arg == "LASTID") {
+          arg = ArgS(args, ++i);
+          // TODO: implement lastID
+          if (!ParseID(arg, true, 0, &opts.lastid)) {
+            return (*cntx)->SendError(kInvalidStreamId, kSyntaxErrType);
+          }
+          continue;
+        }
+      }
+      if (arg == "FORCE") {
+        opts.flags |= kClaimForce;
+      } else if (arg == "JUSTID") {
+        opts.flags |= kClaimJustID;
+      } else {
+        return (*cntx)->SendError("Unknown argument given for XCLAIM command", kSyntaxErr);
+      }
+    }
+  };
+
+  ParseXclaimOptions();
+  if (auto now = GetCurrentTimeMs();
+      opts.deliverytime < 0 || static_cast<uint64_t>(opts.deliverytime) > now)
+    opts.deliverytime = now;
+
+  auto cb = [&](Transaction* t, EngineShard* shard) {
+    return OpClaim(t->GetOpArgs(shard), key, opts, absl::Span{ids.data(), ids.size()});
+  };
+
+  OpResult<ClaimInfo> result = cntx->transaction->ScheduleSingleHopT(std::move(cb));
+  if (result) {
+    ClaimInfo cresult = result.value();
+    if (cresult.justid) {
+      (*cntx)->StartArray(cresult.ids.size());
+      for (size_t i = 0; i < cresult.ids.size(); i++) {
+        (*cntx)->SendBulkString(StreamIdRepr(cresult.ids[i]));
+      }
+    } else {
+      RecordVec crec = cresult.records;
+      (*cntx)->StartArray(crec.size());
+      for (const auto& item : crec) {
+        (*cntx)->StartArray(2);
+        (*cntx)->SendBulkString(StreamIdRepr(item.id));
+        (*cntx)->StartArray(item.kv_arr.size() * 2);
+        for (const auto& k_v : item.kv_arr) {
+          (*cntx)->SendBulkString(k_v.first);
+          (*cntx)->SendBulkString(k_v.second);
+        }
+      }
+    }
+    return;
+  }
+  (*cntx)->SendError(result.status());
```

Comment:
I'd do an early return in case of an error, i.e. change the `if (result) {` in line 1598 to `if (!result) { SendError(); return; }`